src/PostgREST/CLI.hs

```diff
@@ -1,19 +1,72 @@
+{-# LANGUAGE NamedFieldPuns  #-}
+{-# LANGUAGE RecordWildCards #-}
 {-# LANGUAGE QuasiQuotes #-}
 module PostgREST.CLI
-  ( CLI (..)
+  ( main
+  , CLI (..)
   , Command (..)
   , readCLIShowHelp
   ) where
 
 import qualified Options.Applicative as O
 import qualified Protolude.Conv      as Conv
+import qualified Data.Aeson                 as Aeson
+import qualified Data.ByteString.Lazy       as LBS
+import qualified Hasql.Pool                 as P
+import qualified Hasql.Transaction.Sessions as HT
 
 import Text.Heredoc (str)
+import Data.Text.IO (hPutStrLn)
 
 import PostgREST.Version (prettyVersion)
+import PostgREST.AppState    (AppState)
+import PostgREST.Config      (AppConfig (..), dumpAppConfig,
+                              readAppConfig)
+import PostgREST.DbStructure (getDbStructure, getPgVersion)
+import PostgREST.Workers     (reReadConfig)
 
-import Protolude
+import qualified PostgREST.App      as App
+import qualified PostgREST.AppState as AppState
 
+import Protolude hiding (hPutStrLn)
+
+
+main :: App.SignalHandlerInstaller -> App.SocketRunner -> CLI -> IO ()
+main installSignalHandlers runAppInSocket CLI{cliCommand, cliPath} = do
+  conf@AppConfig{..} <- either panic identity <$> readAppConfig mempty cliPath Nothing
+  appState <- AppState.init conf
+
+  -- Override the config with config options from the db
+  -- TODO: the same operation is repeated on connectionWorker, ideally this
+  -- would be done only once, but dump CmdDumpConfig needs it for tests.
+  when configDbConfig $ reReadConfig True appState
+
+  run cliCommand appState
+  where
+    run :: Command -> AppState -> IO ()
+    run CmdDumpConfig appState = putStr . dumpAppConfig =<< AppState.config appState
+    run CmdDumpSchema appState = putStrLn =<< dumpSchema appState
+    run CmdRun appState = App.run installSignalHandlers runAppInSocket appState
+
+-- | Dump DbStructure schema to JSON
+dumpSchema :: AppState -> IO LBS.ByteString
+dumpSchema appState = do
```

Comment:
Similar to how `dumpAppConfig` is in `Config` - this would belong in `DbStructure`, right?

Reply:
Hm, part of it maybe, e.g. the encode function. Other than with dumpConfig, this is an impure function with error handling that is specific to this use case. That felt like it fits best with CLI..

Actually, most of it would not be needed if we loaded the dbStructure synchronously on startup, before launching the web server. That would mean that we bind to the port/socket a bit later. But no request can succeed anyway before we have the dbStructure. Could be a slightly breaking change... @steve-chavez is there a strong reason why we load the dbStructure on a separate thread? It's clear why we do that on reloads, but on startup?

Reply:
> is there a strong reason why we load the dbStructure on a separate thread? It's clear why we do that on reloads, but on startup?

Yes. The db connection can fail at startup. For example, the db can be ready on another machine but DNS resolution can fail(and postgrest errs). While connection recovery happens, it's useful for the web server to reply with a 503 so its health can be monitored. This is important when having many postgrest instances on production.

> dumpSchema appState 

Would it be possible to remove the `dumpSchema` and instead rely on the `request.spec`(https://github.com/PostgREST/postgrest/pull/1794) for debugging? Seems both would give the same information.

Reply:
> Would it be possible to remove the `dumpSchema` and instead rely on the `request.spec`(#1794) for debugging? Seems both would give the same information.

-1. `dumpSchema` is a much more low-level view on things - ideally. Like right now it should not be filtered at all, while `request.spec` already is. Down the road, the `request.spec` stuff might not even return the full schema cache at all - depending on what we figure out is actually needed for the root endpoint. It'd be better if we didn't limit us in reducing `request.spec` later on - if we use it for debugging, we basically need to expose everything.

Reply:
> Hm, part of it maybe, e.g. the encode function. Other than with dumpConfig, this is an impure function with error handling that is specific to this use case. That felt like it fits best with CLI..

Ah. I thought about the purity argument briefly, too. I agree. I suggest to rename `dumpConfig` then. The analogy between the two functions is just misleading. I think `dump` would be the IO action. Maybe `Config.toText conf` or something would be better.

Reply:
> Yes. The db connection can fail at startup. For example, the db can be ready on another machine but DNS resolution can fail(and postgrest errs). While connection recovery happens, it's useful for the web server to reply with a 503 so its health can be monitored. This is important when having many postgrest instances on production.

Ah. I see your point. However, I wonder how that plays with this:

```haskell
when configDbConfig $ reReadConfig True appState
```

Doesn't this wait on the main thread until the connection to the DB could be made, too? Otherwise `--dump-config` wouldn't really work. Maybe I'm completely off, though, because I don't really understand, yet, how all those workers interact and which parts are synchronous or asynchronous.

Reply:
> when configDbConfig $ reReadConfig True appState
> Doesn't this wait on the main thread until the connection to the DB could be made, too? 

Ah, no. If that one fails, it just prints an error - the connection is not retried there(but the in-db config is retried again later on the connection worker, noted in the comment). Really I added that part so the tests can still pass with the cli dump config.

> -1. dumpSchema is a much more low-level view on things

Really my main motivation to not use `dumpSchema` was so I can merge the in-db config with the schema cache, as we discussed on https://github.com/PostgREST/postgrest/pull/1760#discussion_r589070303. Then rely on endpoints for both dumpSchema/dumpConfig(not sure if possible) and avoid synchronization work. But now I think we should go with the MVar alternative I mentioned on that comment. The CLI would wait for the connection worker to be done.

Reply:
> > when configDbConfig $ reReadConfig True appState
> > Doesn't this wait on the main thread until the connection to the DB could be made, too?
> 
> Ah, no. If that one fails, it just prints an error - the connection is not retried there(but the in-db config is retried again later on the connection worker, noted in the comment). Really I added that part so the tests can still pass with the cli dump config.

Ah, I mis-understood the following bit:

```haskell
      if startingUp then
        panic err -- die on invalid config if the program is starting up
```

This only fails when the config is invalid - but not when the connection failed, right?

Reply:
> But now I think we should go with the MVar alternative I mentioned on that comment. The CLI would wait for the connection worker to be done.

I think after Remo removed the timing bit for `--dump-schema`, it doesn't really matter any more in terms of performance. If we want to have a performance test of our schema queries, we should probably add something like that as a unit test or so.

I have no idea what a MVar is - but would that mean that we'd open the socket/port for `--dump-config` / `--dump-schema`, too? I don't think that would be a good idea.

Reply:
> This only fails when the config is invalid - but not when the connection failed, right?

Yes, correct. I can see that part is convoluted. Once the in-db config is inside the schema cache, that code can be removed.

> I think after Remo removed the timing bit for --dump-schema, it doesn't really matter any more in terms of performance.

Sounds good, the MVar would be viable then.